### ISSUE TITLE
Changing from non-secret to secret does not encrypt existing otherwise-unchanged outputs in the state file

### ISSUE THREAD

**Issue by Jmen** [pulumi/pulumi#3772]
When using .net and there are secrets used for values in the resources. These values are not encrypted in the state when the values is passed to a resource output

This can be seen when using a self managed backend like S3

example program which recreates the issue is here 
https://gist.github.com/Jmen/27f1eaafe3660f384c34b65af6b13d4f

and example of the plain text secret in the resource output of the state is here
https://gist.github.com/Jmen/343468cfe680e12d3613bda9b8378b2b

**Comment by mikhailshilkov** [comment]
Hey @Jmen thanks for reporting this! I copied your program exactly but failed to reproduce the result.

Here is a dump of my state file
https://gist.github.com/mikhailshilkov/2b9f26b30b689341441fbd2ee79aeb4e#file-state-json-L100

I'm on Pulumi CLI 1.8.1, the only NuGet package reference is `<PackageReference Include="Pulumi.Aws" Version="1.18.0-preview" />`, using `pulumi login --local`.

Any ideas what I'm doing wrong?

**Comment by Jmen** [comment]
Right, so this is much more nuanced than I thought, as I couldn't recreate the problem at first either. So I retraced my steps from my command history, and I've found how to recreate the issue consistently

I pushed that program up, but the first time, I made a mistake and I had the plain text value in the code, e.g.

```
var secretParameter = new Parameter("my-secret-value", new ParameterArgs
{
  Type = "SecureString",
  Value = "my secret value"
});
```

Then I corrected the problem using the config.RequireSecret()

```
var secretParameter = new Parameter("my-secret-value", new ParameterArgs
{
  Type = "SecureString",
  Value = config.RequireSecret("secretValue")
});
```

and the second time `pulumi up` is run, there are no updates detected, but the Input of the resource in the state is updated with the encrypted value, and the output is left as the plain text value

This only happens if the value being encrypted on the update is the same as the initial plain text value set. If they are different, then a normal updated is detected, and the encrypted version is written to the input and output of the state as expected. That was a tricky one to find 😄

**Comment by Jmen** [comment]
I'm assuming when the compare is done, it is not taking into account the change of a value to being a secret as a state change?

**Comment by Jmen** [comment]
Just confirmed that the javascript implementation has the same behaviour as above

**Comment by lukehoban** [comment]
Thanks for the detailed repro @Jmen!  That is very helpful.  Sounds like the issue isn't that secrets are being leaked, but that they aren't correctly being *un*-leaked.   Definitely important to address, but gives a much better idea where the underlying issue might be.

**Comment by mikhailshilkov** [comment]
Thanks @Jmen so much! Now I'm able to reproduce the issue with any SDK.

I've spent some time digging into the issue and here is what I have so far:
0. We change an input of an AWS resource from some string `"abc"` to `pulumi.secret("abc")`.
1. The engine calls the AWS provider to calculate the diff.
2. Because the provider doesn't understand secrets natively (`AcceptSecrets` not set to `true`), the engine encodes secrets as plain strings.
3. The old and the new values become equal as far as the provider is concerned, so the provider returns `DiffNone`.
4. Since there are no changes, the state is kept intact.
5. We have a now-secret value in the state file.

I posted a naive fix to this in #3779. I suspect there might be side effects to this, which I don't know of. @pgavlin @lukehoban could you take a look and identify what a proper fix might be?

**Comment by mikhailshilkov** [comment]
I'll remove P1 since we now know this is a much more narrow scenario than the original issue. I'll watch for @pgavlin ideas on this diff-ing issue.

**Comment by pgavlin** [comment]
> Since there are no changes, the state is kept intact.

We should probably update the inputs on the resource's state in this case. I believe that those inputs will then be written to the state file.

**Comment by mikhailshilkov** [comment]
@pgavlin I need a bit more hand-holding here... 

> We should probably update the inputs on the resource's state in this case

That was my thought process as well in https://github.com/pulumi/pulumi/pull/3779/files: if `oldInputs` and `newInputs` are not equal, mark it as a diff. However, you rightfully noticed that

> I don't think that this is quite the right fix--I suspect that this will cause unintentional updates. I'll have to stare at the code a bit to figure out what exactly we ought to do.

So how do we update the inputs without causing a diff?

**Comment by pgavlin** [comment]
Hm. Oddly enough, this seems like it should already be handled: https://github.com/pulumi/pulumi/blob/master/pkg/resource/deploy/step_generator.go#L293-L314

In particular, line 313 should update the inputs, which should then be written to the state file.

I wonder if the problem is actually that the engine isn't properly propagating secret-ness for checked inputs when used with providers that do not support secrets.

**Comment by mikhailshilkov** [comment]
Do we re-write to the state file when there's no diff in a resource?

**Comment by mikhailshilkov** [comment]
Actually we do overwrite inputs, the problem is with outputs. 

I just ran this test:
1. Create a `Parameter` with an insecure value and another resource.
2. Run `pulumi up -y` and export the state. It contains the plain text value (OK).
3. Change the parameter value to the same string but secure. Also change another resource and run `pulumi up -y`.
4. Export the state which shows
```
"id": "my-secret-value-6f8240d",
"type": "aws:ssm/parameter:Parameter",
"inputs": {
    "__defaults": [
        "name",
        "tier"
    ],
    "name": "my-secret-value-6f8240d",
    "tier": "Standard",
    "type": "SecureString",
    "value": {
        "4dabf18193072939515e22adb298388d": "1b47061264138c4ac30d75fd1eb44270",
        "ciphertext": "AAABADbb32amaxphxvJZ0EdVStDsD4PfdXJYHVIMISeery5nhw=="
    }
},
"outputs": {
    "allowedPattern": "",
    "arn": "arn:aws:ssm:eu-central-1:153052954103:parameter/my-secret-value-6f8240d",
    "description": "",
    "id": "my-secret-value-6f8240d",
    "keyId": "alias/aws/ssm",
    "name": "my-secret-value-6f8240d",
    "tags": {},
    "tier": "Standard",
    "type": "SecureString",
    "value": "123",
    "version": 1
},
```

So we do overwrite inputs but not outputs. Two problems:

- Outputs are still not encrypted after the update
- The resource displays no change, so a) it's confusing b) if the secretness is the only change in the stack, a user might not run the update at all
